frontend/test/metabase/lib/expressions/generator.js

```diff
@@ -167,11 +95,411 @@ export function generateExpression(seed) {
     if (str === null) {
       throw new Error(`Unknown AST node ${type}`);
     }
-    return str;
+    return String(str);
   };
 
-  let depth = 17;
+  function generateTokenSeq() {
+    const literal = () => {
+      const exp = () => randomItem(["", "-", "+"]) + randomInt(1e2);
+      const number = () => oneOf([zero, one, integer, float1, float2])();
+      const sci = () => number() + randomItem(["e", "E"]) + exp();
+      return {
+        type: NODE.Literal,
+        value: oneOf([number, sci, string])(),
+      };
+    };
+
+    const field = () => {
+      const fk = () => "[" + identifier() + " → " + identifier() + "]";
+      const bracketedName = () => "[" + identifier() + "]";
+      const name = oneOf([identifier, fk, bracketedName])();
+      return {
+        type: NODE.Field,
+        value: name,
+      };
+    };
+
+    const unary = () => {
+      return {
+        type: NODE.Unary,
+        op: randomItem(["-", "NOT "]),
+        child: expression(),
+      };
+    };
+
+    const binary = () => {
+      return {
+        type: NODE.Binary,
+        op: randomItem([
+          "+",
+          "-",
+          "*",
+          "/",
+          "=",
+          "!=",
+          "<",
+          ">",
+          "<=",
+          ">=",
+          " AND ",
+          " OR ",
+        ]),
+        left: expression(),
+        right: expression(),
+      };
+    };
+
+    const call = () => {
+      const count = randomInt(5);
+      return {
+        type: NODE.FunctionCall,
+        value: identifier(),
+        params: listOf(count, [expression])(),
+      };
+    };
+
+    const group = () => {
+      return {
+        type: NODE.Group,
+        child: primary(),
+      };
+    };
+
+    const primary = () => {
+      --depth;
+      const node = oneOf([field, literal, unary, binary, call, group])();
+      ++depth;
+      return node;
+    };
+    const expression = () => (depth <= 0 ? literal() : primary());
+
+    const format = node => {
```

Comment:
Isn't this the exact same format implementation as above?

Reply:
Good catch, thanks! I've since remove this duplicated function.